CIP XXX Marlowe Oracle Protocol/v0.1.0-alpha.md

```diff
@@ -0,0 +1,252 @@
+
+<img src="./diagrams/utxo-marlowe-data-publishing.svg" alt="Data Publishing Using Reference Inputs" style="width: 60%" />
+
+At the end the data are provided to the consumers through the datum in similar fashion to the publishing feeds but they won't include any extra signature so authenticity of the data has to be proven using some other means which we discuss below.
+
+
+##### Cross-Marlowe Choice Sharing
+
+> **FIXME**: Should we drop this section and move it to a separate MIP and discussion page? Should we link this separate discussion from there?
+
+In the current version of the Marlowe validator it is impossible to use choice value from the other instance of Marlowe. We could imagine that on the validator level choice value is read from another Marlowe contract UTxOs as described above but in such a case the choice name which is used to identify that choice should uniquely identify the query in a cross-contract fashion. This is exactly the nature of oracle choices described in this document.
+It seems that in order to properly implement such a feature we should introduce to the language itself a notion of a `ChoiceId` which is "globally" unique. To do so we probably have to introduce `ChoiceName` to the language:
+
+
+```haskell
+data ChoiceName = LocalChoiceName String | GloballyUniqueChoiceName String
+data ChoiceId = ChoiceId ChoiceName Party
+```
+
+##### Authenticity Of The Data
+
+###### Forging Marlowe Choice
+
+On the UTxO level Marlowe validator ensures that the transaction which delivers the choice value was signed by a key corresponding to the party from the contract. After that step every Marlowe thread UTxO protects and keeps that choice together with the party information and value in a map in the datum.
+
+  <img src="./diagrams/utxo-marlowe-choice-verification.svg" alt="On-Chain Marlowe Choice Verification" style="width: 45%" />
+
+Even though Marlowe validator manages that information carefully so it is copied from the inputs to the transaction outputs we can not really be sure that the original information was actually verified in the first place. On the Cardano blockchain it is possible to create an arbitrary output. Malicious actor Eve could easily fake previously presented output just by publishing a transaction with precisely crafted output:
+
+  <img src="./diagrams/utxo-marlowe-choice-fake.svg" alt="Eve Is Publishing Fake Marlowe Choice" style="width: 45%" />
+
+The above scheme is possible because the transaction which Eve constructed does not involve Marlowe validator execution at all! Additionally the choice itself which is stored in the datum is not paired with the data provider signature so she can actually create an arbitrary data point and pair it with arbitrary `Party` (represented by public key hash in our case).
+
+###### Thread Token
+
+In order to fix the above problem we could significantly change the validator and require an extra signature to be provided and stored together with the data point.
```

Comment:
Since we don't follow this suggestion, let's move it, either to the end of the section, or to an appendix.

Reply:
Oh, thread token is actually important from the whole data sharing perspective - I'm leaving that section Simon.